File: rust/520.detect-capital.rs
/*
 * @lc app=leetcode id=520 lang=rust
 *
 * [520] Detect Capital
 */

// @lc code=start
impl Solution {
    pub fn detect_capital_use(word: String) -> bool {
        if word.len() == 1 {
            return true;
        }
        let word = word.chars().collect::<Vec<_>>();
        let first_is_lowercase = word[0].is_lowercase();
        let second_is_lowercase = word[1].is_lowercase();
        if first_is_lowercase && !second_is_lowercase {
            return false;
        }
        for i in 2..word.len() {
            if word[i].is_lowercase() != second_is_lowercase {
                return false;
            }
        }
        true
    }
}
// @lc code=end

File: javascript/520.detect-capital.js
/*
 * @lc app=leetcode id=520 lang=javascript
 *
 * [520] Detect Capital
 */

// @lc code=start
/**
 * @param {string} word
 * @return {boolean}
 */
var detectCapitalUse = function (word) {
  if (word.length === 1) return true;

  // Solution 1:
  // const d = "A".charCodeAt();
  // const firstIsLowercase = word[0].charCodeAt() - d > 26;
  // const secondIsLowercase = word[1].charCodeAt() - d > 26;
  // if (firstIsLowercase && !secondIsLowercase) return false;
  // for (let i = 2; i < word.length; i++) {
  //   if (word[i].charCodeAt() - d > 26 !== secondIsLowercase) return false;
  // }
  // return true;

  // Solution 2:
  return (
    word === word.toUpperCase() ||
    word === word[0] + word.substring(1).toLowerCase()
  );
};
// @lc code=end
